src/main/java/daomephsta/unpick/api/ConstantUninliner.java

```diff
@@ -95,42 +128,98 @@ else if (insn instanceof InsnNode && insn.getOpcode() >= Opcodes.IRETURN && insn
-			}
+			Consumer<Context> ret = processParameterSource(methodOwner, method, parameterSource);
+			if (ret != null)
+				return ret;
 		}
+		for (UnpickValue.MethodUsage methodUsage : unpickValue.getMethodUsages())
+		{
+			Consumer<Context> ret = processMethodUsage(methodUsage);
+			if (ret != null)
+				return ret;
+		}
+		for (AbstractInsnNode usage : unpickValue.getUsages())
+		{
+			Consumer<Context> ret = processUsage(methodOwner, method, usage);
+			if (ret != null)
+				return ret;
+		}
+
+		return null;
 	}
 
-	private void processMethodReturn(String methodOwner, MethodNode method, AbstractInsnNode returnInsn, ReplacementSet replacementSet, 
-			Frame<SourceValue>[] frames, int instructionIndex)
+	private Consumer<Context> processParameterSource(String methodOwner, MethodNode enclosingMethod, int parameterIndex)
 	{
-		if (mapper.targets(methodOwner, method.name, method.desc) 
-				&& mapper.targetsReturn(methodOwner, method.name, method.desc))
+		if (!mapper.targets(methodOwner, enclosingMethod.name, enclosingMethod.desc))
+			return null;
+		if (!mapper.targetsParameter(methodOwner, enclosingMethod.name, enclosingMethod.desc, parameterIndex))
+			return null;
+		logger.log(Level.INFO, String.format("Using enclosing method %s.%s%s parameter %d", methodOwner, enclosingMethod.name, enclosingMethod.desc, parameterIndex));
+		return context -> mapper.mapParameter(methodOwner, enclosingMethod.name, enclosingMethod.desc, parameterIndex, context);
+	}
+
+	private Consumer<Context> processMethodUsage(UnpickValue.MethodUsage methodUsage)
+	{
+		if (methodUsage.getMethodInvocation().getOpcode() == Opcodes.INVOKEDYNAMIC)
 		{
-			Frame<SourceValue> frame = frames[instructionIndex];
-			for (AbstractInsnNode sourceInsn : frame.getStack(0).insns) //Only one value on the stack before a return
+			InvokeDynamicInsnNode invokeDynamicInsn = (InvokeDynamicInsnNode) methodUsage.getMethodInvocation();
+
+			if ("java/lang/invoke/LambdaMetafactory".equals(invokeDynamicInsn.bsm.getOwner()) && "metafactory".equals(invokeDynamicInsn.bsm.getName()))
 			{
-				logger.log(Level.INFO, String.format("Considering target: %s.%s#return Arg Seed: %s", 
-						methodOwner, method.name, Utils.visitableToString(sourceInsn::accept).trim()));
-				Context context = new Context(constantResolver, replacementSet, sourceInsn, method.instructions, frames, logger);
-				mapper.mapReturn(methodOwner, method.name, method.desc, context );
+				Handle lambdaMethod = (Handle) invokeDynamicInsn.bsmArgs[1];
+				if (!mapper.targets(lambdaMethod.getOwner(), lambdaMethod.getName(), lambdaMethod.getDesc()))
+					return null;
+				int kind = lambdaMethod.getTag();
+				boolean hasThis = kind != Opcodes.H_GETSTATIC && kind != Opcodes.H_PUTSTATIC && kind != Opcodes.H_INVOKESTATIC && kind != Opcodes.H_NEWINVOKESPECIAL;
+				int paramIndex = hasThis ? methodUsage.getParamIndex() - 1 : methodUsage.getParamIndex();
```

Comment:
This looks fishy in combination with https://github.com/FabricMC/unpick/pull/2/files#diff-3bfe0a42b980f531f55fb4459eee45e2640babbb87228728fb8db519939fd11cR166 - potentially correcting for `this` twice

Reply:
It's not duplicated logic. The `invokedynamic` instruction has two method invocations when used for lambdas. The main one takes the captured arguments as its parameters and returns the SAM type. Therefore, there is no `this` for the `invokedynamic` instruction itself. Then, `LambdaMetafactory.metafactory` takes a `MethodHandle` for which (usually synthetic, but not always in the case of method references) method to call. This method call may or may not be a static method, so needs to be corrected for accordingly.
This is a special case for when the BSM is `LambdaMetafactory.metafactory`, and handles captured lambda arguments from the enclosing scope. It won't apply for when the BSM is something else, and I didn't really want to put this special case in the middle of the interpreter, thought it would be cleaner inside `ConstantUninliner`.

Reply:
Is it possible to track the slot in the unpick method info, like what's done in enigma's mapping format? I don't see why we need a corrected param index that has to offset for instance operations

Reply:
Unfortunately, unpick doesn't include `this` in its parameter format.

Reply:
Ok thanks for checking again, I didn't look too closely, yet didn't want to leave this unverified